### PR TITLE
Adjust 'this person can edit this widget' checks on My Widgets page, locks widget instances in creator.

### LINKED ISSUE #48: Lock widgets in the creator instead of at the My Widgets page.
See https://github.com/ucfopen/Materia/issues/1202

### PR DESCRIPTION
**Pull request by FrenjaminBanklin**
Closes #48.

Relies on #47.

Detaches the act of locking a widget instance from clicking the 'Edit Widget' button in the My Widgets page and does it in the creator instead.

Only remaining (possible) issue: If a student tries getting clever and goes straight to the create URL of a widget instance, the checks will now prevent the creator from loading - but will still lock that instance.

Either need to consider retooling how the various methods work, or create a new endpoint that can unlock a widget instance.

### CODE REVIEW

File: src/js/controllers/ctrl-create.js [hunk 1]
```diff
@@ -542,7 +556,9 @@ ${msg.toLowerCase()}`,
 		getQset().then(() => {
 			if (!$scope.invalid) {
 				$q(resolve => resolve(inst_id))
+					.then(widgetSrv.lockWidget)
 					.then(widgetSrv.getWidget)
```
Comment (iturgeon) [inline src/js/controllers/ctrl-create.js]: I'm thinking we should try to do lockWidget, getWidget, and checkUserPublishPerms in parallel?

File: src/js/controllers/ctrl-selectedwidget.js [hunk 2]
```diff
@@ -54,21 +54,28 @@ app.controller('SelectedWidgetController', function(
...
-					} else {
-						$scope.show.editPublishedWarning = true
-					}
+	const _editWidgetPromise = () => {
+		return Materia.Coms.Json.send('widget_instance_edit_perms_verify', [
+			$scope.selected.widget.id
+		]).then(response => {
+			if (response.is_locked) {
+				$scope.alert.msg =
+					'This widget is currently locked, you will be able to edit this widget when it is no longer being edited by somebody else.'
```
Comment (iturgeon) [inline src/js/controllers/ctrl-selectedwidget.js]: I think we probably shouldn't use `locked` with the user
```suggestion
					'Someone else is editing this widget, you will be able to edit after they finish.'
```

File: src/js/controllers/ctrl-selectedwidget.test.js [hunk 3]
```diff
@@ -67,4 +69,35 @@ describe('adminWidgetController', () => {
...
+
+		Namespace('Materia.Coms.Json').send = jest.fn().mockResolvedValueOnce({
+			is_locked: true,
+			can_publish: true
+		})
+
+		return $scope.jestTest._editWidgetPromise().then(() => {
+			expect(Materia.Coms.Json.send).toHaveBeenCalledWith('widget_instance_edit_perms_verify', [1])
+			expect($scope.alert.msg).toBe(
+				'This widget is currently locked, you will be able to edit this widget when it is no longer being edited by somebody else.'
```
Comment (iturgeon) [inline src/js/controllers/ctrl-selectedwidget.test.js]: Update to match change above.
```suggestion
				'Someone else is editing this widget, you will be able to edit after they finish.'
```

File: src/js/services/srv-widget.js [hunk 4]
```diff
@@ -53,6 +53,20 @@ app.service('widgetSrv', function(selectedWidgetSrv, dateTimeServ, $q, $rootScop
...
 	}
 
+	const lockWidget = (id = null) => {
+		const deferred = $q.defer()
+		Materia.Coms.Json.send('widget_instance_lock', [id]).then(success => {
+			if (success) {
+				deferred.resolve(id)
+			} else {
+				deferred.reject(
+					'This widget is currently locked, you will be able to edit this widget when it is no longer being edited by somebody else.'
```
Comment (iturgeon) [inline src/js/services/srv-widget.js]: This exact string is in ctrl-my-widgets.js.  Can we de-duplicate it?

File: src/js/services/srv-widget.test.js [hunk 5]
```diff
@@ -401,4 +402,49 @@ describe('widgetSrv', () => {
...
+		let promiseCatch = jest.fn()
+		_service
+			.lockWidget(1)
+			.then(promiseSpy)
+			.catch(promiseCatch)
+		$scope.$digest() // processes promise
+
+		expect(Materia.Coms.Json.send).toHaveBeenCalledWith('widget_instance_lock', [1])
+		expect(promiseSpy).not.toHaveBeenCalled()
+		expect(promiseCatch).toHaveBeenCalledWith(
```
Comment (iturgeon) [inline src/js/services/srv-widget.test.js]: Interesting approach I hadn't thought of.  I think it's ok as is, however one shouldn't have to test that promiseSpy wasn't called - if promiseCatch ends up getting called promiseSpy will always be skipped.

FWIW jest has another way to test catching with `expect().rejects` https://jestjs.io/docs/en/tutorial-async#rejects.  Though I realize this is using angular's promise lib - which may not be compatible

File: src/js/controllers/ctrl-create.js (hunk 1 above)
Comment (FrenjaminBanklin) [reply src/js/controllers/ctrl-create.js]: I'm not sure - if the widget is locked then we shouldn't try to get it, and if we can't get it then we shouldn't bother checking whether the user can publish it. That, and `checkUserPublishPerms` relies on the full widget data being passed into it from `widgetSrv.getWidget`.

File: src/js/controllers/ctrl-selectedwidget.js (hunk 2 above)
Comment (FrenjaminBanklin) [reply src/js/controllers/ctrl-selectedwidget.js]: Feels nicer. I'll swap it in.

**Pull request merged**